HeartbeatLogging/Tests/Unit/HeartbeatStorageTests.swift

```diff
@@ -87,201 +87,218 @@ class HeartbeatStorageTests: XCTestCase {
+  func testGetAndSet_ReturnsOldValueAndSetsNewValue() throws {
     // Given
     let expectation = expectation(description: #function)
-    expectation.expectedFulfillmentCount = 4
+    let heartbeatStorage = HeartbeatStorage(id: #function, storage: StorageFake())
+
+    var dummyHeartbeatInfo = HeartbeatInfo(capacity: 1)
+    dummyHeartbeatInfo.append(Heartbeat(agent: "dummy_agent", date: Date()))
+
+    // When
+    XCTAssertNoThrow(
+      try heartbeatStorage.getAndSet { heartbeatInfo in
+        // Assert that heartbeat storage is empty.
+        XCTAssertNil(heartbeatInfo)
+        // Write new value.
+        return dummyHeartbeatInfo
+      }
+    )
+
+    // Then
+    XCTAssertNoThrow(
+      try heartbeatStorage.getAndSet { heartbeatInfo in
+        expectation.fulfill()
+        // Assert old value is read.
+        XCTAssertEqual(
+          heartbeatInfo?.makeHeartbeatsPayload(),
+          dummyHeartbeatInfo.makeHeartbeatsPayload()
+        )
+        // Write some new value.
+        return heartbeatInfo
+      }
+    )
+
+    wait(for: [expectation], timeout: 0.5)
+  }
+
+  func testGetAndSet_WhenLoadFails_PassesNilToBlockAndReturnsNil() throws {
+    // Given
+    let expectation = expectation(description: #function)
+    expectation.expectedFulfillmentCount = 2
 
     let storageFake = StorageFake()
-    let heartbeatStorage = HeartbeatStorage(id: #file, storage: storageFake)
+    let heartbeatStorage = HeartbeatStorage(id: #function, storage: storageFake)
 
     // When
     storageFake.onRead = {
-      expectation.fulfill() // Fulfilled 2 times.
-      throw DummyError.error
+      expectation.fulfill()
+      return try XCTUnwrap("BAD_DATA".data(using: .utf8))
     }
 
     // Then
-    heartbeatStorage.readAndWriteAsync { heartbeatInfo in
-      expectation.fulfill() // Fulfilled 1 time.
-      XCTAssertNil(heartbeatInfo)
-      return nil
-    }
-
-    try heartbeatStorage.getAndReset { heartbeatInfo in
-      expectation.fulfill() // Fulfilled 1 time.
+    try heartbeatStorage.getAndSet { heartbeatInfo in
+      expectation.fulfill()
       XCTAssertNil(heartbeatInfo)
-      return nil
+      return heartbeatInfo
     }
 
     wait(for: [expectation], timeout: 0.5)
   }
 
-  func testReadAndWriteThenGetAndSet() {
-    let storage = HeartbeatStorage(id: #file, storage: StorageFake())
-
+  func testGetAndSet_WhenSaveFails_ThrowsError() throws {
+    // Given
     let expectation = expectation(description: #function)
+    let storageFake = StorageFake()
+    let heartbeatStorage = HeartbeatStorage(id: #function, storage: storageFake)
 
-    storage.readAndWriteAsync { heartbeatInfo in
-      XCTAssertNil(heartbeatInfo)
+    // When
+    storageFake.onWrite = { _ in
       expectation.fulfill()
-      return heartbeatInfo
+      throw StorageError.writeError
     }
 
+    // Then
+    XCTAssertThrowsError(
+      try heartbeatStorage.getAndSet { heartbeatInfo in
+        heartbeatInfo
+      }
```

Comment:
Nit / in case you wanted to make this shorter, you can use `try heartbeatStorage.getAndSet { $0 }` to do the same thing (using implicitly named arguments). Totally fair if you wanted to be explicit here, but just in case you weren't aware!

Reply:
Reworked in 0c827b651

Reply:
Reworked in 0c827b6